### ISSUE TITLE
How to use a custom unit? Confused.

### ISSUE THREAD

**Issue by rBatt** [ropensci/EML#129]
First, note that on the front page of the repo, the read me's link "[define a custom unit](https://github.com/ropensci/EML#)" simply returns one back to the same page – this is it's own issue perhaps.

When I do `?eml`, I see the `custom_units` argument, and that has a link to the help file for `eml_define_unit`. As far as I can tell I've successfully created the custom unit, but I can't seem to use it properly:

```R

# create custom unit for "count per unit effort (cntpue)"
cntpue.unit <- eml_define_unit("numberPerHectare", unitType="otherUnitType", parentSI="numberPerKilometerSquared", multiplierToSI="0.01", description="number per hectare")

```

I'll create a reproducible example below. These data are bottom trawl data from the Aleutian Islands. I'm trying to follow the instructions from the package readme, and from: https://github.com/ropensci/EML/blob/master/vignettes/Advanced_writing_of_EML.Rmd

```R
library(data.table)
library(EML)

# head of the data set I'm using
ai.mdData <- structure(list(year = c(1983L, 1983L, 1983L, 1983L, 1983L, 1983L, 
1983L, 1983L, 1983L, 1983L), datetime = c("07/15/1983 18:00", 
"07/15/1983 18:00", "07/15/1983 18:00", "07/15/1983 18:00", "07/15/1983 18:00", 
"07/15/1983 18:00", "07/15/1983 18:00", "07/15/1983 18:00", "07/15/1983 18:00", 
"07/15/1983 18:00"), spp = c("Bathymaster signatus", "Berryteuthis magister", 
"Icelus spiniger", "Malacocottus zonurus", "Neptunea sp.", "Pleurogrammus monopterygius", 
"Rajidae", "Sebastes alutus", "Sebastes babcocki", "Sebastes polyspinis"
), haulid = c("554-198301-  1", "554-198301-  1", "554-198301-  1", 
"554-198301-  1", "554-198301-  1", "554-198301-  1", "554-198301-  1", 
"554-198301-  1", "554-198301-  1", "554-198301-  1"), stratum = c("-189.5 52.5", 
"-189.5 52.5", "-189.5 52.5", "-189.5 52.5", "-189.5 52.5", "-189.5 52.5", 
"-189.5 52.5", "-189.5 52.5", "-189.5 52.5", "-189.5 52.5"), 
    stratumarea = c(4065L, 4065L, 4065L, 4065L, 4065L, 4065L, 
    4065L, 4065L, 4065L, 4065L), lat = c(52.91667, 52.91667, 
    52.91667, 52.91667, 52.91667, 52.91667, 52.91667, 52.91667, 
    52.91667, 52.91667), lon = c(-189.18667, -189.18667, -189.18667, 
    -189.18667, -189.18667, -189.18667, -189.18667, -189.18667, 
    -189.18667, -189.18667), depth = c(181L, 181L, 181L, 181L, 
    181L, 181L, 181L, 181L, 181L, 181L), stemp = c(7.4, 7.4, 
    7.4, 7.4, 7.4, 7.4, 7.4, 7.4, 7.4, 7.4), btemp = c(4, 4, 
    4, 4, 4, 4, 4, 4, 4, 4), wtcpue = c(0.5629, 6.0319, 0.7988, 
    2.865, 0.022, 791.9208, 0.7642, 257.8379, 0.0817, 22.9767
    ), cntcpue = c(1.7333, NA, 5.824, 27.2479, NA, 2644.1603, 
    0.0693, 674.0574, 0.0693, 53.4559), region = c("AFSC_Aleutians", 
    "AFSC_Aleutians", "AFSC_Aleutians", "AFSC_Aleutians", "AFSC_Aleutians", 
    "AFSC_Aleutians", "AFSC_Aleutians", "AFSC_Aleutians", "AFSC_Aleutians", 
    "AFSC_Aleutians"), s.reg = c("ai", "ai", "ai", "ai", "ai", 
    "ai", "ai", "ai", "ai", "ai")), .Names = c("year", "datetime", 
"spp", "haulid", "stratum", "stratumarea", "lat", "lon", "depth", 
"stemp", "btemp", "wtcpue", "cntcpue", "region", "s.reg"), sorted = c("year", 
"datetime", "spp", "haulid", "stratum", "stratumarea", "lat", 
"lon", "depth"), class = c("data.table", "data.frame"), row.names = c(NA, 
-10L)) # note I removed the dput() output part giving the internal pointer location

# Create column definitions
ai.md.col.defs <- c(
	"year" = "year of haul", 
	"datetime" = "the day and time of the haul", 
	"spp" = "species scientific name; Genus species", 
	"haulid" = "a unique identifier for the haul; vessel ID - cruise ID - haul number", 
	"stratum" = "the center of the 1 degree grid cell in which the haul was taken; longitude latitude", 
	"stratumarea" = "the area of the statistical stratum (no longer in data set)", 
	"lat" = "latitude of the haul", 
	"lon" = "longitude of the haul, in western hemisphere degrees (for lon > 0, do lon-360)", 
	"depth" = "the maximum depth of the water at the location of the haul", 
	"stemp" = "water temperature at the surface at the location of the haul", 
	"btemp" = "water temperature at the bottom at the location of the haul",
	"wtcpue" = "weight (mass) of the catch", 
	"cntcpue"="number of individuals caught per hectare in the haul", 
	"region" = "the name of this data set, identified by north american sampling region and by the organization doing the sampling",
	"s.reg" = "short region name"
)

# Create custom cntpue unit (duplicates what I have near the start of this issue)
# need numberPerHectare for cntcpue
cntpue.unit <- eml_define_unit("numberPerHectare", unitType="otherUnitType", parentSI="numberPerKilometerSquared", multiplierToSI="0.01", description="number per hectare")

# Create the unit definitions (note this is likely where I'm screwing up, and I've tried a few things here)
ai.md.unit.defs <- list(
	"year"=c(
		unit = "number",
		precision = 1
		), 
	"datetime" = c(
		format = "MM/DD/YYYY HH:MM"
		), 
	"spp" = "Genus species", 
	"haulid" = "haul id", 
	"stratum" = "1 degree lon lat", 
	"stratumarea"=c(
		unit = "kilometer",
		precision = 1
		), 
	"lat"=c(
		unit = "degree",
		precision = 0.5
		), 
	"lon"=c(
		unit = "degree",
		precision = 0.5
		), 
	"depth"=c(
		unit = "meter",
		precision = 1
		), 
	"stemp"=c(
		unit = "celsius",
		precision = 0.1
		), 
	"btemp"=c(
		unit = "celsius",
		precision = 0.1
		), 
	"wtcpue" =c(
		unit = "kilogramsPerHectare",
		precision = 0.0001
		),
	"cntcpue"=cntpue.unit, # PROBLEMATIC LINE? Also, I know "cntcpue" has an extra c
	"region"="region", 
	"s.reg"="short region"
	
)


# Create a data object associated with meta data (?)
# Gives error: Error in unit.def[[1]] : this S4 class is not subsettable
eml_dataTable(
	ai.mdData,
	col.defs=ai.md.col.defs,
	unit.defs=ai.md.unit.defs,
	description="test meta data for AI",
	filename="testMD.csv"
)


# at the problematic line, I've also tried
# "cntcpue"=c(
# 		unit = "numberPerHectare",
# 		precision = 0.0001
# 		)
# but if I use that, then the eml_dataTable() doesn't seem to recognize the unit definition, as it's asking me to enter more information.

```

It's also possible that I'm not using `eml_dataTable` properly, but I've gotten similar error messages using `eml`.

I'm just trying to write the meta data file, I'm not concerned with writing the actual data set. I don't know if the intended uses would change any of these steps – I'm still trying to figure out how I should be using the package, so sorry for any ignorance.

@cboettig

**Comment by rBatt** [comment]
@cboettig 

Here is a simpler example that produces the same problem for me:

```R
test <- data.frame("idt"=1:10, "valt"=rnorm(10))

new.unit <- eml_define_unit("numberPerHectare", unitType="otherUnitType", parentSI="numberPerKilometerSquared", multiplierToSI="0.01", description="number per hectare")

col.defs <- c(
		"idt"="some ID",
		"valt"="some value"
)
unit.defs <- list(
		"idt"=c(
			unit = "number",
			precision = 1
		),
)

blah <- eml(dat=test, title="testMD", col.defs=col.defs, unit.defs=unit.defs, creator="Ryan Batt <battrd@gmail.com>", custom_units=new.unit)
```

**Comment by cboettig** [comment]
To be fixed via #110
